### ISSUE TITLE
Event sources definition in serverless.yaml file

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#1324]
We need to define a great to use API for the event sources definition of the `serverless.yaml` file.

It should be easy to read, remember and extend (as we might want to add features to this later on).

# Proposals
Here are some proposal for the syntax definition:

## Current implementation
```
events:
  aws:
    s3:
      - first-bucket
    http_endpoints:
      post: users/create
   schedule: rate(10 minutes)
```

Upsides:
- Reduces LOC to write
- Prefixed provider

Downsides:
- Syntax is not consistent / hard to remember
- Hard to add features later on as each plugin might need to parse the input differently

---

## Proposal 1
```
events:
  - type: aws:scheduled
    rate: 10 minutes

  - type: aws:s3
    bucket: mybucket

  - type: aws:s3
    bucket: mybucket2
     event: alternativeBucketEvent

  - type: aws:httpendpoint
    path: users
    method: post
    cors: true
    caching: true
    authorizer: arn:lambda:…
```

Upsides:
- Reoccurring schema / easy to remember
- Easier to parse (as it always has key / value pairs)
- Easier to extend with additional features

Downsides:
- No provider nesting / separation
- User cannot see the provider directly
- File might get too long

---

## Proposal 2
```
events:
  - aws_scheduled:
      rate: 10 minutes
  - aws_s3:
      bucket: mybucket
  - aws_s3:
      bucket: mybucket2
      event: alternativeBucketEvent
  - aws_http_endpoint:
      path: users
      method: post
      cors: true
      caching: true
      authorizer: arn:lambda:…
```

Upsides:
- Reoccurring schema / easy to remember
- Easier to parse (as it always has key / value pairs)
- Easier to extend with additional features

Downsides:
- No provider nesting / separation
- File might get too long
- Verbose / repetitive provider namespacing

---

## Proposal 3
```
events:
  aws:
    - scheduled:
        rate: 10 minutes
    - s3:
        bucket: mybucket
    - s3:
        bucket: mybucket2
        event: alternativeBucketEvent
    - http_endpoint:
        path: users
        method: post
        cors: true
        caching: true
        authorizer: arn:lambda:…
```

*similar to proposal 2 but namespaced*

Upsides:
- Reoccurring schema / easy to remember
- Easier to parse (as it always has key / value pairs)
- Easier to extend with additional features

Downsides:
- File might get too long

---

## Proposal 4
```
events:
  aws:
    - scheduled:
        rate: 10 minutes
    - s3: mybucket //format: <name>
    - s3:
        bucket: mybucket2
        event: alternativeBucketEvent
    - http_endpoint: POST /user/add //format: <method> <path>
```

Upsides:
- Having both. Simple string definitions and more complex object definitions

Downsides:

**Comment by mthenw** [comment]
One more that I think is worth to consider. It's very similar to 2. and 3. but instead of accepting only object as a event description, it would also accept string in specified format (depends on event type):

```
events:
  aws:
    - scheduled:
        rate: 10 minutes
    - s3: mybucket //format: <name>
    - s3:
        bucket: mybucket2
        event: alternativeBucketEvent
    - http_endpoint: POST /user/add //format: <method> <path>
```

In most cases people will use string variant but there is still a way to change more config options.
Assuming that this will be stated in docs I think it should not provide confusion.

**Comment by flomotlik** [comment]
100% agree with @mthenw proposal. Having string as a basic choice and the option to go into an object configuration makes it easy to understand and remember. Having to remember to always add ```bucket: ```` to every s3 bucket I want to configure notifications for means I have to remember details I might not need at that particular moment. And it easily allows to add more methods in the future while being very short and succinct.

If you want to add several events of a specific type adding them as a list also make sense (although people need to remember that, but we can easily show that in the docs)

**Comment by ac360** [comment]
### Current Implementation:

* This is an inconsistent API.  Strings, arrays or objects can be values.  It's difficult to parse for both humans and computers.
* This design prevents us from adding extra properties which people may request or properties that may be introduced in these services at a later stage.  100% non-agile and treacherous, since we cannot change this syntax without breaking everything.

### Proposal 1:

* Open to this, but need clarity on whether that YAML syntax is correct.
* Some events will have a `type` attribute, which will cause conflicts.  For example, an API Gateway endpoint has a `type` attribute.  It's a common param name.  Maybe `event` could work instead, but feels redundant.  Ultimately, this exercise seems unnecessary since this can be handled without a key in Proposal 2.

### Proposal 2:

* A clean, shallow list of events, that is super easy to comprehend
* Clear namespaces for events, which we can easily document via some type of event directory.  The event namespaces seem like they will integrate nicely into the insights platform as well.  For example, they could be used as a tag to help describe and retrieve functions.
* The YAML syntax in this one is accurate.

### Proposal 3:  

* An interesting middle-ground.  Nice, @mthenw.
* Personally, I don't favor having the string or object option.  It's just more complexity to learn/maintain, and it's more than is needed, imo.  These event arrays are likely going to be short 90% of the time, since most people will only set 1 or 2 events to trigger a function to execute.  The learning/maintenance cost of the string complexity doesn't seem worth shaving off a few extra lines of code.  Lastly, nothing beats consistency in expected values.  Also, we could always add this later.
* This is very subjective, but the provider key feels unnecessary to me.  Writing code to parse the prefix in **Proposal 2** is trivial, and the benefits you get without the provider key are:
     * A clean, shallow list of events, that is easy to comprehend
     * We establish namespaces for events, which we can easily document via some type of event directory.  The namespaces seem like they would integrate nicely into the insights platform as well.  For example, they could be used as a tag to help describe and retrieve functions.
     * No nesting in the `events` array.  Again, 90% of the time there will only be 1 or 2 events in here.  Provider keys increase LOC and create an extra tier for what should be a simple, short list

**Proposal 2 is my vote.**

**Comment by nikgraf** [comment]
One (to me important) fact about mixing objects & strings:

Let's start from a beginners perspective:
You go to the getting started, copy a basic s3 example and use it.

```yaml
events:
  aws:
    - s3: mybucket
```

Pretty cool & simple 👏

Then you figure out you actually want to attach an event. It's very subtle, but you not only add one line for adding one property. You actually change the whole structure to this:

```yaml
events:
  aws:
    - s3:
        bucket: mybucket
        event: alternativeBucketEvent
```

Diff example in a larger structure:
![screen shot 2016-06-15 at 11 42 29](https://cloud.githubusercontent.com/assets/223045/16075321/bda06fa6-32ee-11e6-9bc1-f25db933d618.png)

If you had a working s3 bucket you have to make sure the whole thing works not only the new feature …

In this case
```yaml
events:
  aws:
    - s3:
      bucket: mybucket
```

```yaml
events:
  aws:
    - s3:
      bucket: mybucket
      event: alternativeBucketEvent
```

Diff example in a larger structure:
![screen shot 2016-06-15 at 11 43 32](https://cloud.githubusercontent.com/assets/223045/16075327/c728f070-32ee-11e6-8bf5-976701c3a657.png)

In this case it's very clear that you only changed one feature, because you changed one line.
In addition the diff is also more clear with a consistent structure.

Thoughts?

**Comment by pmuens** [comment]
Great discussion here.
I've added @mthenw proposal ("Proposal 4").

I like the idea of one fast way to define the event and the option to do more complex stuff with the help of objects (like @mthenw has shown).

But the mixture between string and object representation is what I'm not favoring. For me as a user I'd like to have one way how I should do this. In an ideal situation I could also map this to other similar definitions which would be the case in Proposal 2 and Proposal 3.

What I don't like about Proposal 2 is the repetitive `aws_` in each key.
I would vote for proposal 3 as it has the best of Proposal 2 and the provider namespacing of Proposal 1.

--> Proposal 3

**Comment by pmuens** [comment]
@nikgraf that's brilliant! Thanks for pointing that out! 🙌

**Comment by pmuens** [comment]
One thing to keep in mind is that the Proposal 2 from @ac360 includes the knowledge of Terraform.
They already had to think their API through and define a way which is best for their product and users. They might have already gone the way (unfortunately I was not able to find a corresponding discussion about their config syntax).

**Comment by flomotlik** [comment]
# Provider Nesting

I've got 3 responses so far on the questionaire and all were in favor of the Nested (grouped) syntax. Here the parts regarding Prefixing/Nesting from 2 of them (Erik Erikson didn't really address it in a comment, but just selected it):

```
Joel Haubold(Trek10): 
I prefer #1 because, I generally dislike using string prefixes to do grouping.
```

```
Frank Schmid (Stashimi):
Syntax #1 clearly shows which provide is used whereas #2 flattens everything and is not very clearly arranged.
A second pro for #1 is, that Serverless could internally check better if an event is supported by the specific provider and that it is easier to move esp. endpoints from one provide to the other (every rename in #2 would be an additional place where errors can occur).
```

I will wait for more answers, but including our internal opinions voiced so far it looks like we're leaning towards nested provider syntax.

# String and Object Notation

Regarding string and object notation configuration I do think that the mix of both string and object provide the best usability as users do not have to remember any sub properties for the simplest of cases (e.g. simple schedule or just one bucket you want to use for triggering an event). Once somebody wants to create a more complex configuration they have to look up the documentation anyway to know which options are supported exactly and then moving from string to object notation should be easy.

I do see a potential issue with people leaving the string notation in there and just adding the options and therefore having a Yaml file that does not follow the yaml standard (string and object notation mixed together at the same time)

Imho we could go forward with implementing the object notation and ignoring the string notation for now. Implementing string notation as a future optimisation after we've tested it with more users as a shorthand is something that we can still do easily and it wouldn't impact our current implementation a lot.

# Repeated events

I think repeated events should be handled by adding the same type multiple times, the same way it is handled in Proposal 3:

```
events:
  aws:
    - scheduled:
        rate: 10 minutes
    - s3:
        bucket: mybucket
    - s3:
        bucket: mybucket2
```

So all in all I'm voting for proposal 3 to be implemented now and evaluating in the future if we want to implement string arguments so we're getting to proposal 4.

**Comment by pmuens** [comment]
Good points @flomotlik 

I like the idea to go with Proposal 3 as it seems like this is the best of both world / a kind of blending.
Also favor that we start with the objects implementation and add single string support later on.

**Comment by pmuens** [comment]
One thing I discovered recently is that the "simpler usage" might limit configurability.

What if you would like to have e.g. two different GET endpoints for one lambda when you define the `http_endpoints`:
```
events:
      aws:
        http_endpoints:
          post: users/create
          get: users
          get: user
```

won't work. Don't know if that's a use case but shows that the current syntax has some limitations.

**Comment by flomotlik** [comment]
Imho this is an uncommon use case and can be accomplished by adding a
second http event. Each http event can have one entry per http type.
Am 16.06.2016 10:47 schrieb "Philipp Muens" <notifications@github.com>:

**Comment by eahefnawy** [comment]
Regarding which yaml syntax is correct when it comes to an array of objects, I just confirmed that both are correct. The JSON to YAML converter I used previously chose to output the ugly syntax, so I jumped into conclusion. Sorry about that.

It seems so far that the majority is leaning towards proposal 3, which I do like, but after remembering that we may need to support custom events, I'm starting to lean forward the provider prefix. I'm all in when it comes to organizing by nesting, but we if we support custom events we'll need something like this: (which I dislike)

```yml
events:
    aws:
        schedule:
    azure:
        http:
    custom:
        user_create:
```

Generally, if we could solve this in another way, I'd go for nesting by provider. Nesting in yaml is not that messy imo.

As for supporting both strings and objects, I'm with @flomotlik and @mthenw as long as it's documented. And yaml comments could help there. I do believe that most users will need simple configurations, specially those who are getting started with the framework. We can just simply show them how to do the advanced stuff while keeping beginners happy. After all advanced users probably won't have an issue with switching between objects/strings.

So in summary, I'd go with the following complete example config:

```yml
events:
  - aws_schedule: rate(10 minutes)
  - aws_s3: bucketName
  - aws_s3:
       bucket: complexBucket
       event: alternativeBucketEvent
  - aws_http:
        path: users
        method: post
        cors: true
        caching: true
        authorizer: arn:lambda:…
  - user_create:
        customProperty: value
```

**Comment by pmuens** [comment]
Closed with https://github.com/serverless/serverless/pull/1352

**Comment by dfloresgonz** [comment]
How to include cors : true for proposal 4?
